apps/dcos_l4lb/src/dcos_l4lb_mesos_poller.erl

```diff
@@ -228,19 +228,47 @@ push_vips(LocalVIPs) ->
 generate_ops(LocalVIPs, VIPs) ->
     AgentIP = dcos_net_dist:nodeip(),
     {Ops, LocalVIPs0} =
-        lists:foldl(
-            fun ({{Key, riak_dt_orswot} = LKey, Backends}, {Acc, LVIPs}) ->
-                {LBackends, LVIPs0} = mtake(Key, LVIPs, []),
-                case generate_backend_ops(AgentIP, LBackends, Backends) of
-                    [] -> {Acc, LVIPs0};
-                    Ops -> {[{update, LKey, {update, Ops}}|Acc], LVIPs0}
-                end
-            end, {[], LocalVIPs}, VIPs),
+        lists:foldl(fun (VIP, {Ops, LVIPs}) ->
+            generate_vip_ops(AgentIP, LVIPs, VIP, Ops)
+        end, {[], LocalVIPs}, VIPs),
     maps:fold(fun (Key, Backends, Acc) ->
         LKey = {Key, riak_dt_orswot},
         [{update, LKey, {add_all, Backends}} | Acc]
     end, Ops, LocalVIPs0).
 
+-spec(generate_vip_ops(inet:ip4_address(), #{key() => [backend()]},
```

Comment:
There are too many cases here. I think it's easier to read if only one case expression is used. What do you think about these approaches? 

```erlang
-spec(generate_vip_ops(inet:ip4_address(), #{key() => [backend()]},
    {lkey(), [backend()]}, [riak_dt_orswot:orswot_op()]) ->
    {[riak_dt_map:map_field_update()], #{key() => [backend()]}}).
generate_vip_ops(AgentIP, LocalVIPs, VIP, Ops) ->
    {{Key, riak_dt_orswot} = LKey, Backends} = VIP,
    {LocalBackends, LocalVIPs0} = mtake(Key, LocalVIPs, []),
    {PrevLocalBackends, RemoteBackends} =
        lists:partition(fun ({IP, _B}) -> IP =:= AgentIP end, Backends),
    {Added, Removed} =
        dcos_net_utils:complement(LocalBackends, PrevLocalBackends),
    AddOps = [{add_all, Added} || Added =/= []],
    RemoveOps = [{remove_all, Removed} || Removed =/= []],
    case {LocalBackends, RemoteBackends, AddOps ++ RemoveOps} of
        {[], [], _BackendOps} ->
            {[{remove, LKey} | Ops], LocalVIPs0};
        {LocalBackends, RemoteBackends, []} ->
            {Ops, LocalVIPs0};
        {LocalBackends, RemoteBackends, BackendOps} ->
            {[{update, LKey, {update, BackendOps}} | Ops], LocalVIPs0}
    end.
```

or

```erlang
-spec(generate_vip_ops(inet:ip4_address(), #{key() => [backend()]},
    {lkey(), [backend()]}, [riak_dt_orswot:orswot_op()]) ->
    {[riak_dt_map:map_field_update()], #{key() => [backend()]}}).
generate_vip_ops(AgentIP, LocalVIPs, VIP, Ops) ->
    {{Key, riak_dt_orswot} = LKey, Backends} = VIP,
    {LocalBackends, LocalVIPs0} = mtake(Key, LocalVIPs, []),
    case generate_backend_ops(AgentIP, LocalBackends, Backends) of
        {[] = _BackendOps, [] = _Backends0} ->
            {[{remove, LKey} | Ops], LocalVIPs0};
        {[] = _BackendOps, _Backends0} ->
            {Ops, LocalVIPs0};
        {BackendOps, _Backends} ->
            {[{update, LKey, {update, BackendOps}} | Ops], LocalVIPs0}
    end.

-spec(generate_backend_ops(inet:ip4_address(), [backend()], [backend()]) ->
    {[riak_dt_orswot:orswot_op()], [backend()]}).
generate_backend_ops(AgentIP, LocalBackends, Backends) ->
    {PrevLocalBackends, RemoteBackends} =
        lists:partition(fun ({IP, _B}) -> IP =:= AgentIP end, Backends),
    {Added, Removed} =
        dcos_net_utils:complement(LocalBackends, PrevLocalBackends),
    AddOps = [{add_all, Added} || Added =/= []],
    RemoveOps = [{remove_all, Removed} || Removed =/= []],
    {AddOps ++ RemoveOps, RemoteBackends ++ LocalBackends}.
```

Reply:
Thanks for the suggestion. In fact, the result of my first attempt to adjust this piece of code was more in line with what you suggested here, though not as dense. And it is later, when I polished the change, I introduced `vip_should_be_pruned/3` because the code reads more like plain English, so that many months into the future it won't be necessary to decipher these words in plain English from the list operations. Also, the initial version looked like

```erlang
-spec(vip_should_be_pruned(inet:ip4_address(), #{key() => [backend()]},
    {lkey(), [backend()]}) -> boolean()).
vip_should_be_pruned(AgentIP, LocalVIPs, VIP) ->
    {{Key, riak_dt_orswot}, Backends} = VIP,
    NoLocalBackends = not maps:is_key(Key, LocalVIPs),
    RemoteBackends = [B || {IP, _} = B <- Backends, IP =/= AgentIP],
    case {NoLocalBackends, RemoteBackends} of
        {true, []} -> true;
        _ -> false
    end.
```

but I changed it to the current version due to my habit of deferring calculations to when their results are required.

Now getting back to your two suggestions.

1. I like it. It is dense and doesn't sacrifice readability by and large. The `case` statement is a bit overloaded to my taste, so I'd use it with a comment for the first branch saying that we prune the entire VIP since there are no backends for it anymore. Do you concur?
2. In the second suggestion `generate_backend_ops/3` now does something additionally, that is not reflected in its name.

Reply:
Did the 1st option + a comment. Please take a look. Thank you.